ambry-account/src/main/java/com/github/ambry/account/HelixAccountService.java

```diff
@@ -0,0 +1,523 @@
+   *   {@link Account}s are one-to-one mapped. An {@code AccountInfoMap} guarantees no duplicated account
+   *   id or name, nor conflict among the {@link Account}s within it.
+   * </p>
+   * <p>
+   *   Based on the properties, a {@code AccountInfoMap} internally builds index for {@link Account}s using both
+   *   {@link Account}'s id and name as key.
+   * </p>
+   */
+  private class AccountInfoMap {
+    private final Map<String, Account> nameToAccountMap;
+    private final Map<Short, Account> idToAccountMap;
+
+    /**
+     * Constructor.
+     */
+    private AccountInfoMap() {
+      nameToAccountMap = new HashMap<>();
+      idToAccountMap = new HashMap<>();
+    }
+
+    /**
+     * <p>
+     *   Constructs an {@code AccountInfoMap} from a group of {@link Account}s. The {@link Account}s exists
+     *   in the form of a string-to-string map, where the key is the string form of an {@link Account}'s id,
+     *   and the value is the string form of the {@link Account}'s JSON string.
+     * </p>
+     * <p>
+     *   The source {@link Account}s in the {@code accountMap} may have duplicate account Ids or names. The
+     *   constructor makes the best effort to resolve the conflict. Based on the order when iterating on the
+     *   {@link Account}s, the constructor will: 1) override the previously-added {@link Account} if a latter
+     *   one has the same id; and 2) remove the previously-added {@link Account} if a latter one has the same
+     *   name. Once an {@code AccountInfoMap} is constructed, it is guaranteed there is no duplicate id or name.
+     * </p>
+     * @param accountMap A map of {@link Account}s in the form of (accountIdString, accountJSONString).
+     */
+    private AccountInfoMap(Map<String, String> accountMap) {
+      nameToAccountMap = new HashMap<>();
+      idToAccountMap = new HashMap<>();
+      for (Map.Entry<String, String> entry : accountMap.entrySet()) {
+        String idKey = entry.getKey();
+        String valueString = entry.getValue();
+        Account account;
+        try {
+          JSONObject accountJson = new JSONObject(valueString);
+          if (idKey == null) {
+            logger.error(
+                "Invalid account record when reading accountMap in ZNRecord because either idKey={} or accountJson={}"
+                    + "is null", idKey, accountJson);
+            continue;
+          }
+          account = Account.fromJson(accountJson);
+          if (account.getId() != Short.valueOf(idKey)) {
+            logger.error(
+                "Invalid account record when reading accountMap in ZNRecord because idKey and accountId do not match."
+                    + "idKey={}, accountId={}", idKey, account.getId());
+            continue;
+          }
+        } catch (Exception e) {
+          logger.error("Invalid account record when reading accountMap in ZNRecord. key={}, valueString={}", idKey,
+              valueString, e);
+          continue;
```

Comment:
As discussed offline, the wanted behavior is throw exception during startup if the remote data is corrupted, but will just log error and not update the local copy.